subsys/audio_module/audio_module.c

```diff
@@ -0,0 +1,1065 @@
+			}
+		} else {
+			LOG_DBG("No process function for module %s, discarding input",
+				handle->name);
+		}
+
+		data_fifo_block_free(handle->thread.msg_rx, (void **)&msg_rx);
+	}
+}
+
+/**
+ * @brief  Function for opening a module.
+ */
+int audio_module_open(struct audio_module_parameters *parameters,
+		      struct audio_module_configuration *configuration, char *name,
+		      struct audio_module_context *context, struct audio_module_handle *handle)
+{
+	int ret;
+	k_thread_entry_t thread_entry;
+
+	if (parameters == NULL || configuration == NULL || name == NULL || handle == NULL ||
+	    context == NULL) {
+		LOG_DBG("Parameter is NULL for module open function");
+		return -EINVAL;
+	}
+
+	if (handle->state != AMOD_STATE_UNDEFINED) {
+		LOG_DBG("The module is already open");
+		return -EALREADY;
+	}
+
+	ret = validate_parameters(parameters);
+	if (ret) {
+		LOG_DBG("Invalid parameters for module, ret %d", ret);
+		return ret;
+	}
+
+	memset(handle, 0, sizeof(struct audio_module_handle));
+
+	handle->description->type = parameters->description->type;
+	handle->previous_state = AMOD_STATE_UNDEFINED;
+
+	/* Allocate the context memory */
+	handle->context = context;
+
+	memcpy(handle->name, name, CONFIG_AUDIO_MODULE_NAME_SIZE - 1);
+	if (strlen(name) > CONFIG_AUDIO_MODULE_NAME_SIZE - 1) {
+		LOG_INF("Module's instance name truncated to %s", handle->name);
+	}
+	handle->name[CONFIG_AUDIO_MODULE_NAME_SIZE - 1] = '\0';
+
+	handle->description = parameters->description;
+	memcpy(&handle->thread, &parameters->thread,
+	       sizeof(struct audio_module_thread_configuration));
+
+	if (handle->description->functions->open != NULL) {
+		ret = handle->description->functions->open(
+			(struct audio_module_handle_private *)handle, configuration);
+		if (ret) {
+			LOG_DBG("Failed open call to module %s, ret %d", name, ret);
+			return ret;
+		}
+	}
+
+	if (handle->description->functions->configuration_set != NULL) {
+		ret = handle->description->functions->configuration_set(
+			(struct audio_module_handle_private *)handle, configuration);
+		if (ret) {
+			LOG_DBG("Set configuration for module %s send failed, ret %d", handle->name,
+				ret);
+			return ret;
+		}
+	}
+
+	switch (handle->description->type) {
+	case AMOD_TYPE_INPUT:
+		thread_entry = (k_thread_entry_t)module_thread_input;
+		break;
+
+	case AMOD_TYPE_OUTPUT:
+		thread_entry = (k_thread_entry_t)module_thread_output;
+		break;
+
+	case AMOD_TYPE_PROCESS:
+		thread_entry = (k_thread_entry_t)module_thread_in_out;
+		break;
+
+	default:
+		LOG_DBG("Invalid module type %d for module %s", handle->description->type,
+			handle->name);
+		return -EINVAL;
+	}
+
+	sys_slist_init(&handle->hdl_dest_list);
+	k_mutex_init(&handle->dest_mutex);
+
+	handle->thread_id = k_thread_create(
+		&handle->thread_data, handle->thread.stack, handle->thread.stack_size,
+		(k_thread_entry_t)module_thread_in_out, (void *)handle, NULL, NULL,
+		K_PRIO_PREEMPT(handle->thread.priority), 0, K_FOREVER);
```

Comment:
There should likely be a K_NO_WAIT here, and the ret (handle->thread_id) checked for negative values.

Reply:
Can we discuss?

Reply:
Tested via the k_thread_name_set() below I believe.